File: LeetCode/javascript/lc_0110.ts
/**
 * @file     isBalanced.ts
 * @brief    https://leetcode-cn.com/problems/balanced-binary-tree/submissions/
 * @author   Zhu
 * @date     2021-07-31 13:09
 */


/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * 自底向上，理解递归是一个触底反弹的过程
 */
function isBalanced(root: TreeNode | null): boolean {

    function getTreeHeight(node: TreeNode | null): number {
        if (node) {
            const leftHeight = getTreeHeight(node.left);
            const rightHeight = getTreeHeight(node.right);
            if (leftHeight == -1) return -1;
            if (rightHeight == -1) return -1;
            if (Math.abs(leftHeight - rightHeight) > 1) return -1;
            return 1 + Math.max(leftHeight, rightHeight);
        }
        return 0;
    }

    return getTreeHeight(root) !== -1;
};

File: LeetCode/cpp/lc_0110.cpp
/**
 * @file     lc_0110.cpp
 * @brief    https://leetcode-cn.com/problems/balanced-binary-tree/
 * @author   YongDu
 * @date     2021-09-09
 */
// 递归
class Solution {
public:
  bool isBalanced(TreeNode *root) {
    if (root == nullptr) {
      return true;
    }
    return std::abs(getHight(root->left) - getHight(root->right)) < 2 && isBalanced(root->left) &&
           isBalanced(root->right);
  }

private:
  int getHight(const TreeNode *root) {
    if (root == nullptr) {
      return 0;
    }
    return std::max(getHight(root->left), getHight(root->right)) + 1;
  }
};

// 改进版递归
class Solution {
public:
  bool isBalanced(TreeNode *root) { return getDepth(root) == -1 ? false : true; }

private:
  int getDepth(const TreeNode *root) {
    if (root == nullptr) {
      return 0;
    }
    int leftDepth = getDepth(root->left);
    if (leftDepth == -1) {
      return -1;
    }
    int rightDepth = getDepth(root->right);
    if (rightDepth == -1) {
      return -1;
    }
    return std::abs(leftDepth - rightDepth) > 1 ? -1 : std::max(leftDepth, rightDepth) + 1;
  }
};
